### PR TITLE
[CLIPPER-122] Add get_logs function to clipper_manager and fixes container cleanup

### PR DESCRIPTION
**Pull request by dcrankshaw**
I still need to do some testing with this.

### CODE REVIEW

**Comment by dcrankshaw** [comment]
@Corey-Zumar This is ready for review. It's been tested on my machine. I added a call to `get_clipper_logs()` to the end of the integration test, so you should be able to test the whole thing by running `python integration-tests/light_load_all_functionality.py`

File: management/clipper_manager.py [hunk 1]
```diff
@@ -831,21 +849,98 @@ def add_container(self, model_name, model_version):
...
+                        clipper_label=CLIPPER_DOCKER_LABEL))
                 result = self._execute_root(add_container_cmd)
                 return result.return_code == 0
             else:
                 print("Cannot start containers for externally managed model %s"
                       % model_name)
                 return True
 
+    def get_clipper_logs(self, additional_image_names=None):
+        """Returns true if the provided conda environment is compatible with the container os.
```
Comment (Corey-Zumar) [inline management/clipper_manager.py]: This is the method description for `_conda_env_solvable`. Please replace.

File: examples/tutorial/tutorial_part_two.ipynb [hunk 2]
```diff
@@ -252,7 +252,7 @@
    "metadata": {},
    "outputs": [],
    "source": [
-    "clipper.cleanup()"
+    "clipper.stop_all()"
```
Comment (Corey-Zumar) [inline examples/tutorial/tutorial_part_two.ipynb]: By making this replacement, we've rendered our "Warning" image and description irrelevant as `cleanup()` is no longer called throughout the course of the tutorial. We should probably remove the warning info.

File: management/clipper_manager.py [hunk 3]
```diff
@@ -831,21 +849,98 @@ def add_container(self, model_name, model_version):
...
+        Returns
+        -------
+        list(str)
+            Returns a list of local filenames containing the Docker container log snapshots.
+        """
+        container_ids = self._get_clipper_container_ids()
+        if additional_image_names is not None:
+            for image in additional_image_names:
+                container_ids.extend(
+                    self._get_docker_container_id_from_image(image))
```
Comment (Corey-Zumar) [inline management/clipper_manager.py]: Executes a method that is commented out: `_get_docker_container_id_from_image`

File: management/clipper_manager.py [hunk 4]
```diff
@@ -831,21 +849,98 @@ def add_container(self, model_name, model_version):
...
+        try:
+            os.mkdir(CLIPPER_LOGS_PATH)
+        except OSError:
+            pass
```
Comment (Corey-Zumar) [inline management/clipper_manager.py]: Why do we pass here? If we failed to create the directory we should not continue execution. It makes sense to pass only if the command failed due to a preexisting directory.

File: management/clipper_manager.py [hunk 5]
```diff
@@ -831,21 +849,98 @@ def add_container(self, model_name, model_version):
...
+        """
+        containers = self._execute_root(
+            "docker ps -aq --filter label={clipper_label}".format(
+                clipper_label=CLIPPER_DOCKER_LABEL))
+        ids = [l.strip() for l in containers.split("\n")]
+        print("Clipper container IDS found: %s" % str(ids))
+        return ids
+
+    #
+    #     for l in containers.split("\n"):
```
Comment (Corey-Zumar) [inline management/clipper_manager.py]: Handle commented code - other execution points depend on it

File: management/clipper_manager.py [hunk 6]
```diff
@@ -896,30 +991,41 @@ def set_model_version(self, model_name, model_version, num_containers=0):
         for r in range(num_containers):
             self.add_container(model_name, model_version)
 
-    def stop_all(self):
-        """Stops and removes all Docker containers on the host.
+    def stop_model_containers(self):
```
Comment (Corey-Zumar) [inline management/clipper_manager.py]: Method prototype with no body. Remove.

File: management/clipper_manager.py [hunk 7]
```diff
@@ -896,30 +991,41 @@ def set_model_version(self, model_name, model_version, num_containers=0):
...
 
-        This will stop and remove all Docker containers and images
-        from the host and destroy the Docker network Clipper uses.
         """
+        print("Stopping Clipper and all running models...")
         with hide("output", "warnings", "running"):
-            self.stop_all()
-            self._execute_standard(
-                "rm -rf {model_repo}".format(model_repo=MODEL_REPO))
+            # self._execute_root("docker-compose stop", warn_only=True)
```
Comment (Corey-Zumar) [inline management/clipper_manager.py]: Remove commented code

File: management/clipper_manager.py [hunk 8]
```diff
@@ -896,30 +991,41 @@ def set_model_version(self, model_name, model_version, num_containers=0):
...
+            container_id_str = " ".join(container_ids)
+            self._execute_root(
+                "docker stop {ids}".format(ids=container_id_str),
+                warn_only=True)
             self._execute_root(
-                "docker rmi --force $(docker images -q)", warn_only=True)
-            self._execute_root("docker network rm clipper_nw", warn_only=True)
+                "docker rm {ids}".format(ids=container_id_str), warn_only=True)
+
+    # def cleanup(self):
```
Comment (Corey-Zumar) [inline management/clipper_manager.py]: Remove commented code

**Review by Corey-Zumar — CHANGES_REQUESTED** [review]
`get_clipper_logs()` works properly. Feedback concerns commented code and documentation issues

File: management/clipper_manager.py (hunk 1 above)
Comment (dcrankshaw) [reply management/clipper_manager.py]: Done

File: management/clipper_manager.py (hunk 3 above)
Comment (dcrankshaw) [reply management/clipper_manager.py]: Removed the additional_image_names arg and this code block

File: management/clipper_manager.py (hunk 4 above)
Comment (dcrankshaw) [reply management/clipper_manager.py]: Changed to check if directory exists first

File: management/clipper_manager.py (hunk 5 above)
Comment (dcrankshaw) [reply management/clipper_manager.py]: Deleted and removed callers

File: management/clipper_manager.py (hunk 7 above)
Comment (dcrankshaw) [reply management/clipper_manager.py]: Done

File: management/clipper_manager.py (hunk 8 above)
Comment (dcrankshaw) [reply management/clipper_manager.py]: Done

**Comment by dcrankshaw** [comment]
jenkins test this please

File: management/clipper_manager.py [hunk 9]
```diff
@@ -831,21 +850,59 @@ def add_container(self, model_name, model_version):
...
+        Returns
+        -------
+        list(str)
+            Returns a list of local filenames containing the Docker container log snapshots.
+        """
+        container_ids = self._get_clipper_container_ids()
+        cur_time_logs_path = os.path.join(CLIPPER_LOGS_PATH,
+                                          time.strftime("%Y%m%d-%H%M%S"))
+        if not os.path.exists(cur_time_logs_path):
+            os.makedirs(directory)
```
Comment (Corey-Zumar) [inline management/clipper_manager.py]: this should be `os.makedirs(cur_time_logs_path)`

**Review by Corey-Zumar — CHANGES_REQUESTED** [review]

**Review by Corey-Zumar — APPROVED** [review]
LGTM

**Comment by dcrankshaw** [comment]
jenkins test this please

**Pull request merged**